pkg/conversion/conversion_test.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conversion_test
+
+import (
+	"io/ioutil"
+	"testing"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/testapi"
+	_ "github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta1"
+	_ "github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta2"
+	_ "github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1beta3"
+)
+
+func BenchmarkPodConversion(b *testing.B) {
+	data, err := ioutil.ReadFile("pod_example.json")
+	if err != nil {
+		b.Fatalf("unexpected error while reading file: %v", err)
+	}
+	var pod api.Pod
+	if err := api.Scheme.DecodeInto(data, &pod); err != nil {
+		b.Fatalf("unexpected error decoding pod: %v", err)
+	}
+
+	scheme := api.Scheme.Raw()
+	for i := 0; i < b.N; i++ {
+		versionedObj, err := scheme.ConvertToVersion(&pod, testapi.Version())
+		if err != nil {
+			b.Fatalf("Conversion error: %v", err)
+		}
+		_, err = scheme.ConvertToVersion(versionedObj, scheme.InternalVersion)
```

Comment:
Yeah, I'm never positive - I'm pretty sure gccgo *will* optimize it out, and ultimately I think we want golang to optimize that away if it can.

----- Original Message -----
> > +	data, err := ioutil.ReadFile("pod_example.json")
> > +	if err != nil {
> > +		b.Fatalf("unexpected error while reading file: %v", err)
> > +	}
> > +	var pod api.Pod
> > +	if err := api.Scheme.DecodeInto(data, &pod); err != nil {
> > +		b.Fatalf("unexpected error decoding pod: %v", err)
> > +	}
> > +
> > +	scheme := api.Scheme.Raw()
> > +	for i := 0; i < b.N; i++ {
> > +		versionedObj, err := scheme.ConvertToVersion(&pod, testapi.Version())
> > +		if err != nil {
> > +			b.Fatalf("Conversion error: %v", err)
> > +		}
> > +		_, err = scheme.ConvertToVersion(versionedObj, scheme.InternalVersion)
> 
> that's interesting - but from the:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7107#issuecomment-95208804
> it doesn't seem that compiler optimized it out.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7163/files#r28878916